extensions/amp-carousel/0.1/build-dom.js

```diff
@@ -0,0 +1,356 @@
+    Array.from(
+      element.querySelectorAll(`.${escapeCssSelectorIdent(ClassNames.SLIDE)}`)
+    )
+  );
+  assertDomQueryResults(slidesContainer, slideWrappers, slides);
+  return {slides, slidesContainer, slideWrappers};
+}
+
+/**
+ * Builds the DOM necessary for slidescroll carousel.
+ * @param {!Element} element
+ * @return {{
+ *   prevButton: !HTMLDivElement,
+ *   nextButton: !HTMLDivElement
+ *   container?: !HTMLDivElement
+ *   cells?: !HTMLDivElement[]
+ *   slides?: !HTMLDivElement[]
+ *   slidesContainer?: !HTMLDivElement
+ *   slideWrappers?: !HTMLDivElement[]
+ * }}
+ */
+export function buildDom(element) {
+  const type = getType(element);
+  const slideCount = realChildElements(element).length;
+  const slidesDom =
+    type == 'slides'
+      ? buildSlideScrollCarousel(element)
+      : buildScrollableCarousel(element);
+  const controlsDom = buildCarouselControls(element, slideCount);
+
+  return {...controlsDom, ...slidesDom};
+}
+
+/**
+ * @param {!Element} element
+ * @return {string} The default title to use for the next button.
+ * @param {{index?: string, total?: string}} options - The default title to use for the pevious button.
+ */
+export function getNextButtonTitle(element, options = {}) {
+  const prefix =
+    element.getAttribute('data-next-button-aria-label') ||
+    'Next item in carousel';
+  if (getType(element) === 'scroll') {
+    return prefix;
+  }
+
+  const {index, total} = options;
+  return getVerboseButtonTitle(element, {prefix, index, total});
+}
+
+/**
+ * @param {!Element} element
+ * @param {{index?: string, total?: string}} options - The default title to use for the pevious button.
+ * @return {string} The default title to use for the pevious button.
+ */
+export function getPrevButtonTitle(element, options) {
+  const prefix =
+    element.getAttribute('data-prev-button-aria-label') ||
+    'Previous item in carousel';
+  if (getType(element) === 'scroll') {
+    return prefix;
+  }
+
+  const {index, total} = options;
+  return getVerboseButtonTitle(element, {prefix, index, total});
+}
+
+/**
+ * Returns the title for a next or prev button.
+ * Format: "Next item in carousel (X of Y)".
+ *
+ * @param {*} element
+ * @param {{prefix: string, index: string, total:string}} param1
+ * @return {string}
+ */
+function getVerboseButtonTitle(element, {index, prefix, total}) {
+  /**
+   * A format string for the button label. Should be a string, containing two
+   * placeholders of "%s", where the index and total count will go.
+   * @type {string}
+   */
+  const suffixFormat =
+    element.getAttribute('data-button-count-format') || '(%s of %s)';
+  const suffix = suffixFormat.replace('%s', index).replace('%s', total);
+
+  return `${prefix} ${suffix}`;
+}
+
+/**
+ * @param {!Element} element
+ * @return {'slides' | 'scroll'}
+ */
+export function getType(element) {
+  if (element.getAttribute('type') == 'slides') {
+    return 'slides';
+  }
+  return 'scroll';
```

Comment:
This function is testing `== 'slides!` then conditionally returning `"slides"` or `"scroll"`, but then every callsite is doing its own `getType(element) == "scroll"` check. Maybe just make this `isSlides` or `isScroll`, this function becomes simpler (ie. `element.getAttribute('type`) == 'slides'`) and so does every callsite

Reply:
You're right. Benefit here to previous one was the explicit typechecking. I agree it would be simpler to always check for `slides` instead...